Sources/DNSMessage/Encoding/DNSDecoder.swift

```diff
@@ -0,0 +1,53 @@
+import NIOCore
+
+public struct DNSDecoder {
+    public var buffer: ByteBuffer
+
+    public init(buffer: ByteBuffer) {
+        self.buffer = buffer
+    }
+
+    public mutating func readDNSName() throws -> DNSName {
+        var name = DNSName()
```

Comment:
We could maybe write into an `inout DNSName`, so we can recursively call this function without allocating new DNSNames